zerver/lib/actions.py

```diff
@@ -212,6 +212,21 @@ def bot_owner_user_ids(user_profile: UserProfile) -> Set[int]:
 def realm_user_count(realm: Realm) -> int:
     return UserProfile.objects.filter(realm=realm, is_active=True, is_bot=False).count()
 
+def generate_topic_history_from_db_rows(rows: List[Tuple[str, int]]) -> List[Dict[str, Any]]:
+    canonical_topic_names = set()  # type: Set[str]
+    history = []
+    for (topic_name, max_message_id) in rows:
```

Comment:
We should add a comment here that this algorithm relies on `rows` being reverse-sorted by id.  (We probably should have had such a comment in the original code, but it was a bit more obvious before the extraction that the query was ordered.)

Reply:
Agreed; I'll add that and then merge this first commit.  

Arguably we should just rewrite this function to use a dict mapping (canonical_topic) -> [max_message_id, topic_name] and write the update code to correctly not rely on that assumption, just because it'd make the code more readable.